src/Chapter4.hs

```diff
@@ -489,6 +505,14 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+instance Applicative List where
+  pure :: a -> List a
+  pure a = Cons a Empty
+
+  (<*>) :: List (a -> b) -> List a -> List b
+  Empty <*> _ = Empty
+  _ <*> Empty = Empty
+  Cons f fs <*> Cons x xs = Cons (f x) (fs <*> xs)
```

Comment:
Actually, you've implemented a non-standard `Applicative` instance for the list. Here you apply each function only once for the corresponding item from the second list. But in the task description we say that each function should be applied to each element from the second list.

This is still a correct `Applicative` instance. Turns out that the list has two valid `Applicative` instance. But the list type `[]` in Haskell implements a different instance, that we ask to provide in the task. With this implementation the laws of `Monad` and `Applicative` instances do not agree with each other.

Your implementation of `Applicative` for the standard list `[]` is provided via the `ZipList` newtype:

* https://hackage.haskell.org/package/base-4.14.0.0/docs/Control-Applicative.html#t:ZipList

Newtypes are the standard Haskell way to provide multiple instances for a single type. You can't have multiple instances for the same type, so you optionally choose the default (main) one, and provide others for newtypes.

Reply:
Aha, that makes sense. I misunderstood the problem. I will push an update :)